server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchLookupFieldsPhase.java

```diff
@@ -0,0 +1,227 @@
+     * @param mergeLookupFields if {@code true} then resolved lookup fields will be merged as regular fields to {@link DocumentField}
+     * @param listener          the listener to be called when the lookup completes
+     */
+    public void fetchLookupFields(
+        String clusterAlias,
+        Task parentTask,
+        SearchHits hits,
+        boolean onlyLocalNode,
+        boolean recordFailures,
+        boolean mergeLookupFields,
+        ActionListener<Void> listener
+    ) {
+
+        if (clusterAlias != null && RemoteClusterAware.LOCAL_CLUSTER_GROUP_KEY.equals(clusterAlias) == false) {
+            listener.onResponse(null);
+            return;
+        }
+
+        if (mergeLookupFields) {
+            listener = ActionListener.runBefore(listener, () -> {
+                for (SearchHit hit : hits) {
+                    hit.getDocumentFields().values().forEach(DocumentField::mergeLookupFieldsToFields);
+                }
+            });
+        }
+
+        final Map<LookupKey, LookupField> resolvedFields = new HashMap<>();
+        final Map<LookupKey, List<LookupField>> unresolvedFields = new HashMap<>();
+        for (SearchHit hit : hits) {
+            for (DocumentField docField : hit.getDocumentFields().values()) {
+                for (LookupField lookupField : docField.getLookupFields()) {
+                    final LookupKey key = new LookupKey(lookupField);
+                    if (lookupField.isResolved()) {
+                        resolvedFields.put(key, lookupField);
+                    } else {
+                        unresolvedFields.computeIfAbsent(key, k -> new ArrayList<>()).add(lookupField);
+                    }
+                }
+            }
+        }
+        final List<LookupRequest> lookupRequests = new ArrayList<>();
+        for (Map.Entry<LookupKey, List<LookupField>> e : unresolvedFields.entrySet()) {
+            final LookupField resolved = resolvedFields.get(e.getKey());
+            if (resolved != null) {
+                if (resolved.getFailure() != null) {
+                    e.getValue().forEach(v -> v.setFailure(resolved.getFailure()));
+                } else {
+                    e.getValue().forEach(v -> v.setValues(resolved.getValues()));
+                }
+            } else {
+                lookupRequests.add(new LookupRequest(e.getKey(), e.getValue()));
+            }
+        }
+
+        if (lookupRequests.isEmpty()) {
+            listener.onResponse(null);
+            return;
+        }
+
+        if (allowExpensiveQueries.getAsBoolean() == false) {
+            handleFailure(
+                lookupRequests,
+                new ElasticsearchException(
+                    "cannot be executed against lookup fields while [" + ALLOW_EXPENSIVE_QUERIES.getKey() + "] is set to [false]."
+                )
+            );
+            listener.onResponse(null);
+        }
+
+        final MultiGetRequest multiGetRequest = new MultiGetRequest();
```

Comment:
I used MGET because of its performance: it's faster and uses fewer messages. We can reduce more messages with the optimization that uses local nodes only. We can sync on this.